### PR TITLE
Add module ADR

### PR DESCRIPTION
**Pull request by hickeyma**
Partial #13

### CODE REVIEW

**Review by gabe-l-hart — CHANGES_REQUESTED** [review]
Thanks for putting this all in! This is much more thorough than the stub I'd started, so rather than giving you my half-formed version, I've just added a bunch of thoughts/suggestions here.

File: docs/adrs/001-module.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+class MyModule(caikit.core.ModuleBase):
+    # my impl
+
+```
+
+A module wraps one or more AI algorithms, and provides a crisply defined interface for obtaining predictions by providing one or more data model objects as input and returns exactly one data model object as output.  
+
+## Interface
+
+Currently, each module implements the following APIs, some of which can be considered core (implemented by all moduyles),
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:25]: Fixed

File: docs/adrs/001-module.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+such as `.__init__()`, `.run()` and some which may not be present in all (like `.train()`). Additionally, modules come with some helper methods that are useful in analyzing block performance.
+
+#### `__init__()`
+
+`__init__` is the constructor for the module and should take only in-memory data structures, relevant to the block implementation.
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:30]: Updated

File: docs/adrs/001-module.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+## Interface
+
+Currently, each module implements the following APIs, some of which can be considered core (implemented by all moduyles),
+such as `.__init__()`, `.run()` and some which may not be present in all (like `.train()`). Additionally, modules come with some helper methods that are useful in analyzing block performance.
+
+#### `__init__()`
+
+`__init__` is the constructor for the module and should take only in-memory data structures, relevant to the block implementation.  
+Since python does not have constructor overloading, having a primitive, in-memory constructor allows us to implement multiple ways of instantiating blocks by using classmethods.
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:31]: Updated

File: docs/adrs/001-module.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+import caikit
+
+@caikit.module(
+    "00110203-0405-0607-0809-0a0b02dd0e0f", "MyModule", "0.0.1")
+class MyModule(caikit.core.ModuleBase):
+    # my impl
+
+```
+
+A module wraps one or more AI algorithms, and provides a crisply defined interface for obtaining predictions by providing one or more data model objects as input and returns exactly one data model object as output.
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:21]: Updated

File: docs/adrs/001-module.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    # my impl
+
+```
+
+A module wraps one or more AI algorithms, and provides a crisply defined interface for obtaining predictions by providing one or more data model objects as input and returns exactly one data model object as output.  
+
+## Interface
+
+Currently, each module implements the following APIs, some of which can be considered core (implemented by all moduyles),
+such as `.__init__()`, `.run()` and some which may not be present in all (like `.train()`). Additionally, modules come with some helper methods that are useful in analyzing block performance.
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:26]: Is `run()` now optional or required?

File: docs/adrs/001-module.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+#### `load()`
+
+`load` is a `classmethod` that loads the AI model by reading the `config.yml` and any artifacts into memory and then calls the module class' `__init__` method to create and return a new loaded instance of the model. The structure of `config.yml` is standardized to some extent, with some other pieces which can be specified as per class requirements.
+
+#### `run()`:
+Takes instances of classes defined as [data models](https://github.com/caikit/caikit/blob/main/docs/adrs/010-data-model-definition.md) as inputs and returns a single instance of [data model](https://github.com/caikit/caikit/blob/main/docs/adrs/010-data-model-definition.md) as output.
+
+Different algorithms of the same module type return the same type of output, but may take different kinds of inputs.
+This may feel a little overly prescriptive, but it allows us to keep things organized and also to have a generic server that can swap out different algorithms for a given module type.  
+So, `run` is very much like a `predict` method in other frameworks and we often use them in our verbiage interchangeably.
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:42]: Updated

File: docs/adrs/001-module.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+#### `train()`
+
+`train` is a python `@classmethod` that takes one or more [data stream]() instances as inputs containing training data and, potentially, other necessary arguments.
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:46]: Not sure whether we want a reference to the data stream ADR or not here?

File: docs/adrs/001-module.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+The `train` method creates a new instance of the module, in memory, that can later be saved to disk using `save()`.
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:47]: Updated

File: docs/adrs/001-module.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+Different algorithms of the same module type return the same type of output, but may take different kinds of inputs.
+This may feel a little overly prescriptive, but it allows us to keep things organized and also to have a generic server that can swap out different algorithms for a given module type.  
+So, `run` is very much like a `predict` method in other frameworks and we often use them in our verbiage interchangeably.
+
+#### `train()`
+
+`train` is a python `@classmethod` that takes one or more [data stream]() instances as inputs containing training data and, potentially, other necessary arguments.
+The `train` method creates a new instance of the module, in memory, that can later be saved to disk using `save()`.  
+So, `train` is similar to `fit` in some frameworks. Not all modules will have a `train` implemented.
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:48]: Updated

File: docs/adrs/001-module.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+6. `_extract_pred_set(dataset, pre_process_function)`: Instance method to get predicted results from the dataset's input, after running them against the loaded module.
+
+One can call a module's `.evaluate_quality(dataset_path, preprocess_func)` specifying the test dataset path, and internally the module type's base class methods will be triggered to obtain the gold labels and predicted labels, and perform evaluation to return quality metrics.
+
+
+#### `stream()`
+
+This is a helper method which works in conjunction with `caikit.core.data_model.streams.DataStream` to allow running large sets of data for a given model using streams. Refer to [data stream]() for details.
+
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:102]: Added

**Comment by hickeyma** [comment]
Thanks for the review and feedback @gabe-l-hart. Comments updated and some unresolved that require further feedback.

Ready for review again.

File: docs/adrs/001-module.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (gabe-l-hart) [reply docs/adrs/001-module.md:46]: Yeah, an ADR would be a good link here too. No strong preferences.

**Comment by gabe-l-hart** [comment]
@hickeyma Responded to questions 🏓

File: docs/adrs/001-module.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:26]: Understand now. Updated to specify this.

File: docs/adrs/001-module.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+#### `load()`
+
+`load` is a `classmethod` that loads the AI model by reading the `config.yml` and any artifacts into memory and then calls the module class' `__init__` method to create and return a new loaded instance of the model. The structure of `config.yml` is standardized to some extent, with some other pieces which can be specified as per class requirements.
+
+#### `run()`:
+Takes instances of classes defined as [data models](https://github.com/caikit/caikit/blob/main/docs/adrs/010-data-model-definition.md) as inputs and returns a single instance of [data model](https://github.com/caikit/caikit/blob/main/docs/adrs/010-data-model-definition.md) as output.
+
+Different algorithms of the same module type return the same type of output, but may take different kinds of inputs.
+This may feel a little overly prescriptive, but it allows us to keep things organized and also to have a generic server that can swap out different algorithms for a given module type.
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:41]: Removing the sentences for now. We can add to this later when we have the task ADR.

File: docs/adrs/001-module.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:46]: I'll leave it with the empty reference for now and we can add the add when it is ready

File: docs/adrs/001-module.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+Since python does not have constructor overloading, having a primitive, in-memory constructor allows us to implement multiple ways of instantiating blocks by using classmethods.
+
+#### `load()`
+
+`load` is a `classmethod` that loads the AI model by reading the `config.yml` and any artifacts into memory and then calls the module class' `__init__` method to create and return a new loaded instance of the model. The structure of `config.yml` is standardized to some extent, with some other pieces which can be specified as per class requirements.
+
+#### `run()`:
+Takes instances of classes defined as [data models](https://github.com/caikit/caikit/blob/main/docs/adrs/010-data-model-definition.md) as inputs and returns a single instance of [data model](https://github.com/caikit/caikit/blob/main/docs/adrs/010-data-model-definition.md) as output.
+
+Different algorithms of the same module type return the same type of output, but may take different kinds of inputs.
```
[the remark this answers is not in the document]
Comment (hickeyma) [reply docs/adrs/001-module.md:40]: Removed for now, lets revisit when task add available.

File: docs/adrs/001-module.md (hunk 1 above)
Comment (gabe-l-hart) [inline docs/adrs/001-module.md:25]:
```suggestion
Currently, each module implements the following APIs, some of which can be considered core (implemented by all modules),
```

**Review by gabe-l-hart — CHANGES_REQUESTED** [review]
One more typo NIT, and it looks like stream didn't get removed, even though you commented Done

**Pull request closed without merging**